consensus/tendermint/rule_proposal_and_polka_previous_test.go

```diff
@@ -0,0 +1,97 @@
+package tendermint
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestProposalAndPolkaPrevious(t *testing.T) {
+	t.Run("Line 28: locked value is unchanged", func(t *testing.T) {
+		// In this test, we're validator 1. Validator 0 is faulty.
+		stateMachine := setupStateMachine(t, 4, 3)
+
+		firstRound := newTestRound(t, stateMachine, 0, 0)
+		secondRound := newTestRound(t, stateMachine, 0, 1)
+
+		wrongValue := value(42)
+		correctValue := value(43)
+
+		// In the first round, validator 0 "tricked" us with invalid value.
+		firstRound.start()
+		firstRound.validator(0).proposal(wrongValue, -1).expectActions(
+			firstRound.action().broadcastPrevote(&wrongValue),
+		)
+		firstRound.validator(0).prevote(&correctValue)
+		firstRound.validator(1).prevote(&correctValue)
+		firstRound.validator(2).prevote(&correctValue)
+		firstRound.validator(1).precommit(&correctValue)
+		firstRound.validator(2).precommit(&correctValue)
+		firstRound.processTimeout(precommit)
+
+		assertState(t, stateMachine, height(0), round(1), propose)
+
+		secondRound.validator(1).proposal(correctValue, 0).expectActions(
+			secondRound.action().broadcastPrevote(&correctValue),
+		)
+	})
+
+	t.Run("Line 28: locked value change", func(t *testing.T) {
+		// In this test, we're validator 1. Validator 0 is faulty.
```

Comment:
I'm finidng this test difficult to follow. Can we add a description at the top, something like this?

```suggestion
	t.Run("Line 28: locked value change", func(t *testing.T) {
		// In this test, we're validator 1. Validator 0 is faulty.
		// Round 1: We lock on the valid value. The other validators do not lock on it,
		//          because they receive a different valid value.
		// Round 2: Validator 1 now proposes the different value from the one we locked on. We reject it.
```